dotnet/src/Connectors/Connectors.Memory.AzureCognitiveSearch/AzureCognitiveSearchMemory.cs

```diff
@@ -0,0 +1,316 @@
+        var fieldBuilder = new FieldBuilder();
+        var fields = fieldBuilder.Build(typeof(AzureCognitiveSearchRecord));
+        var newIndex = new SearchIndex(indexName, fields)
+        {
+            SemanticSettings = new SemanticSettings
+            {
+                Configurations =
+                {
+                    new SemanticConfiguration("default", new PrioritizedFields
+                    {
+                        TitleField = new SemanticField { FieldName = "Description" },
+                        ContentFields =
+                        {
+                            new SemanticField { FieldName = "Text" },
+                            new SemanticField { FieldName = "AdditionalMetadata" },
+                        }
+                    })
+                }
+            }
+        };
+
+        return this._adminClient.CreateIndexAsync(newIndex, cancellationToken);
+    }
+
+    private async Task<string> UpsertRecordAsync(
+        string indexName,
+        AzureCognitiveSearchRecord record,
+        CancellationToken cancellationToken = default)
+    {
+        var client = await this.GetSearchClientAsync(indexName, cancellationToken).ConfigureAwait(false);
+
+        Response<IndexDocumentsResult>? result = await client.MergeOrUploadDocumentsAsync(
+            new List<AzureCognitiveSearchRecord> { record },
+            new IndexDocumentsOptions { ThrowOnAnyError = true },
+            cancellationToken).ConfigureAwait(false);
+
+        if (result == null)
+        {
+            throw new AzureCognitiveSearchMemoryException("Memory write returned null");
+        }
+
+        return result.Value.Results[0].Key;
+    }
+
+    private static MemoryRecordMetadata ToMemoryRecordMetadata(AzureCognitiveSearchRecord data)
+    {
+        return new MemoryRecordMetadata(
+            isReference: data.IsReference,
+            id: DecodeId(data.Id),
+            text: data.Text ?? string.Empty,
+            description: data.Description ?? string.Empty,
+            externalSourceName: data.ExternalSourceName,
+            additionalMetadata: data.AdditionalMetadata ?? string.Empty);
+    }
+
+    /// <summary>
+    /// Normalize index name to match ACS rules.
+    /// The method doesn't handle all the error scenarios, leaving it to the service
+    /// to throw an error for edge cases not handled locally.
+    /// </summary>
+    /// <param name="indexName">Value to normalize</param>
+    /// <returns>Normalized name</returns>
+    private static string NormalizeIndexName(string indexName)
+    {
+        if (indexName.Length > 128)
+        {
+            throw new AzureCognitiveSearchMemoryException("The collection name is too long, it cannot exceed 128 chars");
+        }
+
+#pragma warning disable CA1308 // The service expects a lowercase string
+        indexName = indexName.ToLowerInvariant();
+#pragma warning restore CA1308
+
+        return Regex.Replace(indexName.Trim(), "[\\s|\\\\|/|.|_|:]", "-");
```

Comment:
Could you add a comment describing the intent of the regex? For example: replace any whitespace, backslash, forward slash, period, underscore or colon with a dash.